File: Data Structures & Algorithms/subsets/submission-0.js
class Solution {
    /**
     * @param {number[]} nums
     * @return {number[][]}
     */
    subsets(nums) {
        const result = [];
        const currentSubset = [];

        const backtrack = (index) => {
            result.push([...currentSubset]);
            for (let i = index; i < nums.length; i++) {
                currentSubset.push(nums[i]);
                backtrack(i + 1);
                currentSubset.pop();
            }
        };

        backtrack(0);
        return result;
    }
}

File: Data Structures & Algorithms/permutations/submission-0.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        result = []
        n = len(nums)

        def backtrack(current_perm):
            if len(current_perm) == n:
                result.append(current_perm[:])
                return
            
            for num in nums:
                if num not in current_perm:
                    current_perm.append(num)
                    backtrack(current_perm)
                    current_perm.pop()
        backtrack([])
        return result